src/uffd_handler/src/prefaulter.rs

```diff
@@ -21,11 +26,266 @@ impl UffdPrefaulter {
+        // Make sure that the start of the chunk to prefault is within the current region.
+        prefault_start_addr = max(prefault_start_addr, region.mapping.base_host_virt_addr);
+
+        // Page faults must be handled in units of pages. Because `bytes_before` is not constraint
+        // to be a multiple of page size, the start address must be rounded down to page boundary.
+        prefault_start_addr = page_start_of_addr(prefault_start_addr)?;
+
+        // If the start of the memory region has already been brought up into RAM
+        // (if for example a page fault was served by copying in RAM more than the faulting page)
+        // UFFD_COPY will return `EEXIST`. We need to find the first page not brought into memory
+        // after `prefault_start_addr`.
+        // This unwrap is safe because it is guaranteed that the key is the start of a page
+        // within the current region.
+        if *region.page_states.get(&prefault_start_addr).unwrap() == MemPageState::FromFile {
+            // Look for an uninitialized page up to the faulting page, which we know for
+            // sure that is not present in memory.
+            while prefault_start_addr < page_fault_start_addr {
+                prefault_start_addr += page_size;
+                // Found the first Uninitialized occurrence, so prefaulting will start from there.
+                // This unwrap is safe because it is guaranteed that the key is the start of a page
+                // within the current region.
+                if *region.page_states.get(&prefault_start_addr).unwrap()
+                    == MemPageState::Uninitialized
+                {
+                    break;
+                }
+            }
+        }
+
+        Ok(prefault_start_addr)
+    }
+
+    /// Return the end address of the chunk to prefault into memory. The address is obtained
+    /// by adding `after` amount of bytes to the end of the faulting page. If the end address
+    /// exceeds the current region, the chunk is truncated to fit into the memory region
+    /// (the end of the current memory region is returned).
+    /// When there is no customized prefaulting configured, the minimum chunk to copy to RAM
+    /// is the faulting page.
+    pub fn get_prefaulting_end_address(
+        &self,
+        page_fault_start_addr: usize,
+        region: &MemRegion,
+    ) -> Result<usize, HandlerError> {
+        let page_size = get_page_size().map_err(HandlerError::PageSize)?;
+
+        // This will not overflow because it has been verified beforehand during mappings
+        // deserialization.
+        let region_end = region.mapping.base_host_virt_addr + region.mapping.size;
+        // We add `page_size` because we always want to at least bring into RAM the faulting page,
+        // even if customized prefaulting is not configured.
+        let prefault_chunk_end = self
+            .bytes_after
+            .checked_add(page_fault_start_addr + page_size)
```

Comment:
Maybe we could add another comment here that `page_fault_start_addr + page_size` could not overflow (as we know that the maximum value is the size of the region)?